### PR TITLE
fix(js): respect blank lines in parameters

### LINKED ISSUE #5260: Blank lines aren’t preserved between parameters
**Prettier 1.14.3**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgGIQQ7AA6UOOkUmMATgK5gwS0AUZFF0cANB5zn4UYAdwh9ynGAAtacXkJwAzCPVoSBgyZ0VKAlgDcF2nJj0APDZ0xwjUKxTh6A5tJgOcUPVGMD49xS0BOFRbBED+AEpiAF8yOKgyJXooJj1oZUJ2E24PIKkxDxk5X04VNTyIk30jDzNLRRs7DydXd0UvHyKESpMQsMSTLWjgGJAeEAgABxh06mRQdFpaCBEABSWETGQQdAMIPQATcZAAI1p0MABrOBgAZSnL72dkOnpeEG8bWhg1i+cALboZBKdCoGwTABWmHMACELtdbnd0AC4AAZbxwEFgiEgR60b47U7oU4AT1Q0BOU1o3hgAHUjjJkAAOAAME2pEBsdIuUx21Lg3yMJzkAEd6Ho5H90IDgUhQeCPjYAXpXgwlc9QgBFegQeDYxUTGAkhmHJlIABMRouelQzwAwhAAUCdlBuCd6DYACok7bynFwGIxIA)
```sh
--parser babylon
```

**Input:**
```jsx
class Foo {
  constructor(
    one,
     
    two,
    three,
    four,
     
     
    five,
    six,
    seven,
    eight,
    nine,
    ten,
     
    eleven
     
  ) {}
}

function foo(
    one,
     
    two,
    three,
    four,
     
     
    five,
    six,
    seven,
    eight,
    nine,
    ten,
     
    eleven
     
) {}
```

**Output:**
```jsx
class Foo {
  constructor(
    one,
    two,
    three,
    four,
    five,
    six,
    seven,
    eight,
    nine,
    ten,
    eleven
  ) {}
}

function foo(
  one,
  two,
  three,
  four,
  five,
  six,
  seven,
  eight,
  nine,
  ten,
  eleven
) {}

```

**Expected behavior:**

```jsx
class Foo {
  constructor(
    one,

    two,
    three,
    four,

    five,
    six,
    seven,
    eight,
    nine,
    ten,

    eleven
  ) {}
}

function foo(
  one,

  two,
  three,
  four,

  five,
  six,
  seven,
  eight,
  nine,
  ten,

  eleven
) {}

```

Note that the blank lines are preserved when _calling_ a function:

**Prettier 1.14.3**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzCEAUAdKACAvaOAGl0IPMJgHcIz9qALAJzlKoPQFcWGK8nSoy4BLAG4cReAM6iAHv0Iy4kqEoJxRAcyYwNeKKKhSK8dUMHS4AG1UJLuAJQgSICAAcYo6DOSgAQxYWCBoABSCEPxQA8QhRABNXEAAjFgCwAGs4GABlDwzjbWQYFm5SEGMVFhgw9O0AWwDkVACbFTcAKxl5ACF0rJzcgIa4ABljOBa2jpAClmrkVICUgE8baGSPFmMYAHVEmCZkAA4ABjdtiBU99I8l7bhqyWS2AEduUTY6gMbmpFa7QqKgaohKZWBRTsAEVuBB4NMgW4YCsDgkjsgAEzI9KiGxFADCEAaTSWUGIyW4KgAKitooCVABfRlAA)
```sh
--parser babylon
```

**Input:**
```jsx
foo(
    one,
     
    two,
    three,
    four,
     
     
    five,
    six,
    seven,
    eight,
    nine,
    ten,
     
    eleven
     
)
```

**Output:**
```jsx
foo(
  one,

  two,
  three,
  four,

  five,
  six,
  seven,
  eight,
  nine,
  ten,

  eleven
);

```

This is especially important when there are decorators involved:

**Prettier pr-5188**
[Playground link](https://deploy-preview-5188--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAucAbAhgZ0wAgGIQQ7AA6UOOkUmMATgK5gwS0AUZFFAAgJZQBWcJqwBC6WgEoOnAA60eAN3TwctOOgAm0VAE8cAI3FIcASTG0ANGWnc+g4QFkdABVoQwcbCynlZ8pSpqmtp6ALYubh5etMYmTq7unpgsFjjWvrYCQjCsJllMPNAAKjoyngB0AIJQEDAAFnC0RXV8AOY+NjhyispwqupaULo46DX1jc1tsdW1DU0tUK04AD449FAacABmfHAaVr4SxDgAvmRnUCAWIBAyMIXUyKDibgDuzuIImMgg6AoQPA0VxA+lo6DAAGs4DAAMoycFtZB0ehwa58TCNGCudCtULoZBbdCoDHXfiYAAeIjBkOhMPQoTgABldgSiSSQPDaBjaD8YKVPGB5Hdgd1YAB1QH1ZAADgADNc5BAMWKwTIfnJPI0FHBgWoAI70HhqbG4-FIQnE1EgDGhHhIhhWzBtVBwACK9FqOvNbKtMHQ+glGilSAATNc6OgeKg2gBhCChPE-GpQHXXegYor+77ey0nE5AA)
```sh
--parser typescript
```

**Input:**
```tsx
class Foo {
  constructor(
    @inject(Bar)
    private readonly bar: IBar,

    @inject(MyProcessor)
    private readonly myProcessor: IMyProcessor, 

    @inject(InjectionTypes.AnotherThing)

    private readonly anotherThing: IAnotherThing | undefined,
  ) { }
}

```

**Output:**
```tsx
class Foo {
  constructor(
    @inject(Bar)
    private readonly bar: IBar,
    @inject(MyProcessor)
    private readonly myProcessor: IMyProcessor,
    @inject(InjectionTypes.AnotherThing)
    private readonly anotherThing: IAnotherThing | undefined
  ) {}
}

```

The above also raises the question if blank lines should be allowed between decorators. Not sure.

### PR DESCRIPTION
**Pull request by evilebottnawi**
- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If not an internal change) I’ve added my changes to the `CHANGELOG.unreleased.md` file following the template.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

Fixes #5260

### CODE REVIEW

File: src/language-js/printer-estree.js [hunk 1]
```diff
@@ -4154,15 +4181,13 @@ function printFunctionParams(path, print, options, expandArg, printTypeParams) {
   //   b,
   //   c
   // }) {}
-  if (shouldHugArguments(fun)) {
-    return concat([typeParams, "(", join(", ", printed), ")"]);
+  if (isParametersShouldBeHugged) {
```
Comment (evilebottnawi) [inline src/language-js/printer-estree.js]: Not sure about this.
Example:
```js
// Input
call((
  one,

  two,
  three,
  four,


  five,
  six,
  seven,
  eight,
  nine,
  ten,

  eleven

) => {});

// Output
call((one, two, three, four, five, six, seven, eight, nine, ten, eleven) => {});
```

Should we respect newlines here?

File: src/language-js/printer-estree.js [hunk 2]
```diff
@@ -4087,15 +4087,45 @@ function printFunctionTypeParameters(path, options, print) {
 
 function printFunctionParams(path, print, options, expandArg, printTypeParams) {
   const fun = path.getValue();
+  const parent = path.getParentNode();
   const paramsField = fun.parameters ? "parameters" : "params";
+  const isParametersInTestCall = isTestCall(parent);
+  const isParametersShouldBeHugged = shouldHugArguments(fun);
+  const isParametersShouldBeExpended =
```
Comment (j-f1) [inline src/language-js/printer-estree.js]: Should this be `Expanded`? “expended” means [something else](https://www.wordnik.com/words/expend)

Comment (evilebottnawi) [reply src/language-js/printer-estree.js]: @j-f1 oh, yes typo :+1:

Comment (evilebottnawi) [reply src/language-js/printer-estree.js]: @j-f1 done

**Comment by evilebottnawi** [comment]
Somebody can review? :smile:

File: CHANGELOG.unreleased.md [hunk 3]
```diff
@@ -103,3 +103,62 @@ Examples:
...
+    
+  // Output (Prettier master)
+  function foo(
+    one,
+  
+    two,
+    three,
+    four,
+  
+
```
Comment (lipis) [inline CHANGELOG.unreleased.md]: maybe leaving maximum one empty line?

Comment (evilebottnawi) [reply CHANGELOG.unreleased.md]: @lipis yep, we can, i just think it is better explain changes

Comment (evilebottnawi) [reply CHANGELOG.unreleased.md]: fixes, also add snapshot to ensure

File: src/language-js/printer-estree.js (hunk 1 above)
Comment (lipis) [reply src/language-js/printer-estree.js]: If they can fit in one line.. I would say no.

**Comment by suchipi** [comment]
To clarify, this is just respecting *blank* lines, right? Not adding object literal behavior to function parameters?

**Comment by evilebottnawi** [comment]
@suchipi yep, just respect newlines

**Comment by evilebottnawi** [comment]
Guys, somebody can review because I'm tired of just doing rebase :smile:

**Comment by suchipi** [comment]
@evilebottnawi sorry that we didn't review this one sooner

File: src/language-js/printer-estree.js [hunk 4]
```diff
@@ -4087,15 +4087,45 @@ function printFunctionTypeParameters(path, options, print) {
 
 function printFunctionParams(path, print, options, expandArg, printTypeParams) {
   const fun = path.getValue();
+  const parent = path.getParentNode();
   const paramsField = fun.parameters ? "parameters" : "params";
+  const isParametersInTestCall = isTestCall(parent);
+  const isParametersShouldBeHugged = shouldHugArguments(fun);
+  const isParametersShouldBeExpanded =
```
Comment (duailibe) [inline src/language-js/printer-estree.js]: `isParametersShouldBeHugged` -> `shouldHugParameters` ?

Comment (evilebottnawi) [reply src/language-js/printer-estree.js]: /cc @duailibe done, also rename `isParametersShouldBeExpanded` to `shouldExpandParaments`

**Pull request merged**